viewer/src/components/AdvancedLog.vue

```diff
@@ -43,19 +43,70 @@
 </template>
 <script lang="ts">
 import Vue from "vue";
-import { Component, Prop } from "vue-property-decorator";
-import Engine, { LogEntry } from "@engine";
+import {Component} from "vue-property-decorator";
+import {Command, GaiaHex, LogEntry} from "@gaia-project/engine";
+
+type MoveAndEntry = { move: string; entry: LogEntry };
+
+function ownTurn(e: MoveAndEntry): boolean {
+  if (e.move == null) {
+    return true;
+  }
+  const outOfTurn = [Command.ChargePower, Command.BrainStone, Command.ChooseIncome, Command.Decline];
+  for (const command of outOfTurn) {
+    if (e.move.includes(command)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+type RecentHexSet = { set: Set<GaiaHex>; done: (e: MoveAndEntry) => boolean };
+>>>>>>> 8275e5b (highlight recent actions + color of player who took a board action)
 
 @Component({
   computed: {
-    data() {
+    data () {
       return this.$store.state.gaiaViewer.data;
     },
-    history(): Array<{ move: string; entry: LogEntry }> {
+    history (): Array<MoveAndEntry> {
       const ret = [];
       let advancedLogIndex = 0;
       let nextLogEntry = this.data.advancedLog[advancedLogIndex];
 
+      const addRecent = (s: RecentHexSet, log: Array<MoveAndEntry>) => {
+        s.set.clear();
+
+        for (const e of log) {
+          if (s.done(e)) {
+            return;
+          }
+          if (e.move) {
+            const args = this.data.parseMove(e.move).args;
+            const number = args.indexOf("build");
+            if (number >= 0) {
+              const coord = args[number + 2].replace(".", "");
+              const hex = this.data.map.getS(coord);
+              s.set.add(hex);
+            }
+          }
+        }
+      };
+
+      const addRecentMoves = (log: Array<MoveAndEntry>) => {
+        const context = this.$store.state.gaiaViewer.context;
+        const player = this.$store.state.gaiaViewer.player?.index ?? this.data.currentPlayer;
+
+        const recentSets: RecentHexSet[] = [
+          {set: context.highlighted.recentHexes, done: (e: MoveAndEntry) => e.entry.player === player && ownTurn(e)},
+          {set: context.highlighted.currentRoundHexes, done: (e: MoveAndEntry) => e.entry.round != null}
+        ];
+
+        for (const s of recentSets) {
+          addRecent(s, log);
+        }
+      };
```

Comment:
This should not be here.

Maybe make `currentRoundHexes` and `recentHexes` as store [getters](https://vuex.vuejs.org/guide/getters.html).

To get the moves in the history (which you can then use to extract coordinates);

```ts
import { findLastIndex, findLast } from "lodash";

// round moves
const roundEntryIndex = findLastIndex(advancedLog, logItem => !!logItem.round);
const moveIndex = advancedLog.slice(roundEntryIndex+1).find(logItem => !!logItem.move)?.move;
const roundMoves = moveIndex ? moveHistory.slice(moveIndex) : [];

// recent moves
const lastMoveIndex = findLast(advancedLog, logItem => ownTurn(logItem))?.move
const recentMoves = lastMoveIndex ? moveHistory.slice(lastMoveIndex + 1) : []
```

Reply:
thanks - that sounds a lot cleaner.